### PR TITLE
Use Dart 3 class modifiers in `package:jni`

### LINKED ISSUE #334: ☂️  Upgrade to Dart 3
Make use of the new features.

- [x] `jnigen/test/simple_package_test/runtime_test_registrant.dart`: use records instead of map
- [x] Make generated impl classes `abstract interface class`
- [x] Make primitive types final.
- [x] Make `JType` sealed.

More to be added...

### PR DESCRIPTION
**Pull request by HosseinYousefi**
Closes #334.

### CODE REVIEW

File: jni/CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,10 @@
 - **Breaking Change**: The default return `callType` of type parameter `int` for
   methods such as `JObject.callMethodByName<int>` is now Java's `long` instead
   of `int` to be consistent with the way arguments work.
+- **Breaking Change**: `JType` is now `sealed`.
+- **Breaking Change**: Primitive types and their type classes are now `final`.
+- **Breaking Change**: `JArray.filled` now uses the Dart type of the `fill`
```
Comment (dcharkes) [inline jni/CHANGELOG.md:19]: Dart runtime time, Dart static type?

File: jni/lib/src/jarray.dart [hunk 2]
```diff
@@ -92,10 +91,9 @@ class JArray<E> extends JObject {
   /// Creates a [JArray] of the given length with [fill] at each position.
   ///
   /// The [length] must be a non-negative integer.
-  /// The [fill] must be a non-null [JObject].
   static JArray<E> filled<E extends JObject>(int length, E fill) {
-    assert(!fill.isNull, "fill must not be null.");
-    final clazz = fill.getClass();
+    RangeError.checkNotNegative(length);
+    final clazz = fill.$type.getClass();
```
Comment (dcharkes) [inline jni/lib/src/jarray.dart:96]: ArgumentError on `fill.isNull`?

File: jni/lib/src/jprimitives.dart [hunk 3]
```diff
@@ -6,92 +6,92 @@
 // lowercase.
 // ignore_for_file: camel_case_types
 
-import 'types.dart';
+part of 'types.dart';
```
Comment (dcharkes) [inline jni/lib/src/jprimitives.dart:9]: I assume this is needed because the whole class hierarchy needs to be in the same "file"?

**Review by dcharkes — APPROVED** [review]
lgtm! Thanks @HosseinYousefi !

File: jni/lib/src/jarray.dart (hunk 2 above)
Comment (HosseinYousefi) [reply jni/lib/src/jarray.dart:96]: `fill` now can be null.

File: jni/lib/src/jprimitives.dart (hunk 3 above)
Comment (HosseinYousefi) [reply jni/lib/src/jprimitives.dart:9]: Yes! Because of `sealed`.

File: jni/lib/src/jarray.dart (hunk 2 above)
Comment (dcharkes) [reply jni/lib/src/jarray.dart:96]: Right, we use the Dart runtime type now.

**Pull request merged**